graphql_compiler/schema_transformation/restrict_schema.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from copy import copy
+
+from graphql import build_ast_schema
+from graphql.language.visitor import REMOVE, Visitor, visit
+
+from ..ast_manipulation import get_ast_with_non_null_and_list_stripped
+from .utils import SchemaStructureError, get_query_type_name, get_scalar_names
+
+
+def restrict_schema(schema_ast, types_to_keep):
+    """Return new AST containing only a subset of types.
+
+    In addition to all types named in types_to_keep, the query type will also be kept. All user
+    defined scalars and enums will be kept.
+    All property fields will be kept, and only vertex fields that go to kept types will be
+    kept.
+
+    Args:
+        schema_ast: Document, representing the schema we're using to create a new schema with
+                    fewer types. It is not modified by this function
+        types_to_keep: Set[str], the set of names of types that we want to keep in the output
+                       schema
+
+    Returns:
+        Document, representing the schema that is derived from the input schema AST, but only
```

Comment:
Any thoughts on what we'd do with `type_equivalence_hints`? It might be good to include a helper function that is able to transform `type_equivalence_hints` on the unrestricted schema into the hints on the newly-generated restricted schema.